cmd/clusterctl/pkg/client/cluster/template.go

```diff
@@ -0,0 +1,207 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"encoding/base64"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/repository"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TemplateClient has methods to work with templates stored in the cluster/out of the provider repository.
+type TemplateClient interface {
+	// Returns a workload cluster template stored in the given ConfigMap.
+	GetFromConfigMap(configMapNamespace, configMapName, configMapDataKey, targetNamespace string, listVariablesOnly bool) (repository.Template, error)
+
+	// GetFromURL return a workload cluster reading from the given URL.
+	GetFromURL(templateURL, targetNamespace string, listVariablesOnly bool) (repository.Template, error)
+}
+
+// templateClient implements TemplateClient.
+type templateClient struct {
+	proxy               Proxy
+	configClient        config.Client
+	gitHubClientFactory func(configVariablesClient config.VariablesClient) (*github.Client, error)
+}
+
+// ensure templateClient implements TemplateClient.
+var _ TemplateClient = &templateClient{}
+
+// newTemplateClient returns a templateClient.
+func newTemplateClient(proxy Proxy, configClient config.Client) *templateClient {
+	return &templateClient{
+		proxy:               proxy,
+		configClient:        configClient,
+		gitHubClientFactory: getGitHubClient,
+	}
+}
+
+func (t *templateClient) GetFromConfigMap(configMapNamespace, configMapName, configMapDataKey, targetNamespace string, listVariablesOnly bool) (repository.Template, error) {
+	if configMapNamespace == "" {
+		return nil, errors.New("invalid GetFromConfigMap operation: missing configMapNamespace value")
+	}
+	if configMapName == "" {
+		return nil, errors.New("invalid GetFromConfigMap operation: missing configMapName value")
+	}
+
+	c, err := t.proxy.NewClient()
+	if err != nil {
+		return nil, err
+	}
+
+	configMap := &corev1.ConfigMap{}
+	configMapObjectKey := client.ObjectKey{
+		Namespace: configMapNamespace,
+		Name:      configMapName,
+	}
+
+	if err := c.Get(ctx, configMapObjectKey, configMap); err != nil {
+		return nil, errors.Wrapf(err, "error reading ConfigMap %s/%s", configMapNamespace, configMapName)
+	}
+
+	data, ok := configMap.Data[configMapDataKey]
+	if !ok {
+		return nil, errors.Errorf("the ConfigMap %s/%s does not have the %s data key", configMapNamespace, configMapName, configMapDataKey)
```

Comment:
```suggestion
		return nil, errors.Errorf("the ConfigMap %s/%s does not have the %q data key", configMapNamespace, configMapName, configMapDataKey)
```
I came across this error and it wasn't clear that the key had to be "template". I also suggest adding a hint in the msg output to direct the user to use `--from-config-map-key` so that they can better directed to the next steps. But I can also see how the CLI flag would then be part of the library error output which is not ideal.

Reply:
Another suggestion is to use "data" as the default key since that seems to be the default when creating config maps. Unless the use case is for a config map that holds multiple items it.

Reply:
I'm -1 to add a flag name in library error messages.
WRT to the default name, I'm ok with everything get consensus

Reply:
Yeah...I'm fine with not adding the flag name in the library error msg. |
Maybe we can just switch the `%s` to `%q` then, that way the error can look more like this: `Error: the ConfigMap default/foobar does not have the "template" data key`.